smack-websocket/src/main/java/org/jivesoftware/smack/websocket/implementations/okhttp/OkHttpWebsocketImpl.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.websocket.implementations.okhttp;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.packet.StreamClose;
+import org.jivesoftware.smack.packet.TopLevelStreamElement;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.websocket.LoggingInterceptor;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints;
+import org.jivesoftware.smack.websocket.elements.CloseElement;
+import org.jivesoftware.smack.websocket.elements.OpenElement;
+import org.jivesoftware.smack.websocket.implementations.AbstractWebsocket;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.WebSocket;
+import okhttp3.WebSocketListener;
+
+public final class OkHttpWebsocketImpl extends AbstractWebsocket {
+
+    private static final Logger LOGGER = Logger.getLogger(OkHttpWebsocketImpl.class.getName());
+    private static OkHttpClient okHttpClient = null;
+
+    private boolean closeInitiatedByClient = false;
+    private boolean closeInititatedByServer = false;
+    private String openStreamHeader;
+    private final ModularXmppClientToServerConnectionInternal connectionInternal;
+    private WebSocket currentWebsocket;
+    private WebsocketConnectionPhase phase;
+    private Map<WebsocketRemoteConnectionEndpoint, Throwable> failureMap = new HashMap<>();
+
+    public OkHttpWebsocketImpl(ModularXmppClientToServerConnectionInternal connectionInternal,
+            DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints) {
+        this.connectionInternal = connectionInternal;
+
+        if (okHttpClient == null) {
+            // Creates an instance of okHttp client.
+            buildWebsocketClient();
+        }
+    }
+
+    private static void buildWebsocketClient() {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        // Add some mechanism to enable and disable this intercepter.
+        builder.addInterceptor(new LoggingInterceptor());
```

Comment:
This should be conditionally done depending on whether or not the connection has a `SmackDebugger` attached.